core/state/snapshot/conversion.go

```diff
@@ -360,10 +366,10 @@ func generateTrieRoot(db ethdb.KeyValueWriter, it Iterator, account common.Hash,
 	return stop(nil)
 }
 
-func stackTrieGenerate(db ethdb.KeyValueWriter, in chan trieKV, out chan common.Hash) {
-	t := trie.NewStackTrie(db)
+func stackTrieGenerate(db ethdb.KeyValueWriter, kind string, in chan trieKV, out chan common.Hash) {
+	t := zktrie.NewStackTrie(db)
 	for leaf := range in {
-		t.TryUpdate(leaf.key[:], leaf.value)
+		t.TryUpdateWithKind(kind, leaf.key[:], leaf.value)
```

Comment:
I had considered a similar solution, but the underlying zktrie update requires two parameters: vFlag and vPreimage. If we only have bytes, it might be difficult to deduce the value of vFlag backwards.

Reply:
vFlag should be part of the encoded bytes: put the first 4 bytes for vFlag and item number, then appending with the `byte32` obtained from `MarshalFields`, then you can use these as value bytes for presistent. vPreimage is never a field for presistent.

Reply:
It sounds work, but it involves a lot of minor fixes corresponding to the encoding and we should re-test the whole process to ensure its correctness.
I will take note of this modification, and make the changes to the tests once we have reviewed everything together.